pkg/operator/ceph/cluster/osd/replace.go

```diff
@@ -0,0 +1,167 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package osd for the Ceph OSDs.
+package osd
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	OSDReplaceConfigName       = "osd-replace-config"
+	OSDReplaceConfigKey        = "config"
+	OSDStoreUpdateConfirmation = "yes-really-update-store"
+)
+
+// OSDReplaceInfo represents an OSD that needs to replaced
+type OSDReplaceInfo struct {
+	ID   int    `json:"id"`
+	Path string `json:"path"`
+	Node string `json:"node"`
+}
+
+func (o *OSDReplaceInfo) saveAsConfig(context *clusterd.Context, clusterInfo *cephclient.ClusterInfo) error {
+	configStr, err := o.string()
+	if err != nil {
+		return errors.Wrapf(err, "failed to convert osd replace config to string")
+	}
+
+	newConfigMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      OSDReplaceConfigName,
+			Namespace: clusterInfo.Namespace,
+		},
+		Data: map[string]string{
+			OSDReplaceConfigKey: configStr,
+		},
+	}
+
+	err = clusterInfo.OwnerInfo.SetControllerReference(newConfigMap)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set owner reference on %q configMap", newConfigMap.Name)
+	}
+
+	_, err = k8sutil.CreateOrUpdateConfigMap(clusterInfo.Context, context.Clientset, newConfigMap)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create or update %q configMap", newConfigMap.Name)
+	}
+
+	return nil
+}
+
+func (o *OSDReplaceInfo) string() (string, error) {
+	configInBytes, err := json.Marshal(o)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to marshal osd replace config")
+	}
+
+	return string(configInBytes), nil
+}
+
+// getOSDReplaceInfo returns an existing OSD that needs to be replaced for a new backend store
+func (c *Cluster) getOSDReplaceInfo() (*OSDReplaceInfo, error) {
+	osdReplaceInfo, err := GetOSDReplaceConfigMap(c.context, c.clusterInfo)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get any existing OSD in replace configmap")
+	}
+
+	if osdReplaceInfo != nil {
+		return osdReplaceInfo, nil
+	}
+
+	pgHealthMsg, pgClean, err := cephclient.IsClusterClean(c.context, c.clusterInfo)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to check if the pgs are clean before replacing OSDs")
+	}
+	if !pgClean {
+		logger.Warningf("skipping OSD replacement because pgs are not healthy. PG status: %q", pgHealthMsg)
+		return nil, nil
```

Comment:
Does the reconcile get re-queued? It seems like it will just think there are no OSDs to replace, and complete the reconcile without error.

Reply:
>> It seems like it will just think there are no OSDs to replace, and complete the reconcile without error.
Agreed. We need to way to re-queue here. If we simply return at error here, it will prevent reconcile of other OSDs to reconcile. So any upgade/create operation will be blocked until pgs are back up.

Another approach is the main `reconcile` method. If the `desired` osd store is `bluestore-rdr` and `status.storage.osd.storetype["bluestore"] != 0`, then we reconcile. 

What do you think?

Reply:
You already have a check for PGs clean that waits for ~15 minutes after replacing an OSD. Can we also trigger that wait at the end of the reconcile if a replacement is still needed, but a replacement didn't happen during that specific reconcile? (Or is it already doing this?) Then we don't need to worry about re-queueing here, it can just be requeued after that wait.

Reply:
>> Can we also trigger that wait at the end of the reconcile if a replacement is still needed, but a replacement didn't happen during that specific reconcile? (Or is it already doing this?)

You meant triggering a reconcile, right?

Reply:
Why would we need a `wait` at the end of the reconcile if a replacement is still needed? What exactly will we be waiting for?

Reply:
Returning an error at the end of OSD reconcile when a OSD replace operation was performed. This will mean operator will reconcile every time we replace an OSD. It will also mean operator will reconcile once after all the OSDs are replaced.